src/file_tree.ml

```diff
@@ -127,77 +169,90 @@ type t = Dir.t
-    match Int.compare a.ino b.ino with
-    | Eq -> Int.compare a.dev b.dev
-    | ne -> ne
-
-  let dummy = { ino = 0; dev = 0 }
-
-  let of_stats (st : Unix.stats) =
-    { ino = st.st_ino
-    ; dev = st.st_dev
-    }
-
-  module Map = Map.Make(struct type nonrec t = t let compare = compare end)
-end
-
 let is_temp_file fn =
   String.is_prefix fn ~prefix:".#"
   || String.is_suffix fn ~suffix:".swp"
   || String.is_suffix fn ~suffix:"~"
 
 let load ?(warn_when_seeing_jbuild_file=true) path =
-  let rec walk path ~dirs_visited ~project ~data_only : Dir.t =
-    let contents = lazy (
-      let files, unfiltered_sub_dirs =
-        Path.readdir_unsorted path
-        |> List.filter_partition_map ~f:(fun fn ->
-          let path = Path.relative path fn in
-          if Path.is_in_build_dir path then
-            Skip
-          else begin
-            let is_directory, file =
-              match Unix.stat (Path.to_string path) with
-              | exception _ -> (false, File.dummy)
-              | { st_kind = S_DIR; _ } as st ->
-                (true, File.of_stats st)
-              | _ ->
-                (false, File.dummy)
-            in
-            if is_directory then
-              Right (fn, path, file)
-            else if is_temp_file fn then
+  let open Result.O in
+  let rec walk path ~dirs_visited ~project:parent_project ~data_only : Dir.t =
+    let load_status, files, unfiltered_sub_dirs =
+      let combined = lazy (
+        match Path.readdir_unsorted path with
+        | exception exn ->
+          Errors.warn Loc.none
+            "Unable to read directory %a@. Ignoring. \
+             Remove this message by ignoring by adding:@.\
+             (dirs \\ %s)@.\
+             to %a@.\
+             Reason: %a@."
+            Path.pp path
+            (Path.basename path)
+            Path.pp (Path.relative (Path.parent_exn path) "dune")
+            Exn.pp exn;
+          Error exn
+        | unsorted_contents ->
+          List.filter_partition_map unsorted_contents ~f:(fun fn ->
+            let path = Path.relative path fn in
+            if Path.is_in_build_dir path then
               Skip
-            else
-              Left fn
-          end)
-      in
-      let files = String.Set.of_list files in
-      let unfiltered_sub_dirs =
-        List.sort
-          ~compare:(fun (a, _, _) (b, _, _) -> String.compare a b)
-          unfiltered_sub_dirs
+            else begin
+              let is_directory, file =
+                match Path.stat path with
+                | exception _ -> (false, File.dummy)
+                | { st_kind = S_DIR; _ } as st -> (true, File.of_stats st)
+                | _ -> (false, File.dummy)
+              in
+              if is_directory then
+                Right (fn, path, file)
+              else if is_temp_file fn then
+                Skip
+              else
+                Left fn
+            end)
+          |> Result.ok
+      ) in
+      let load_status = lazy (
```

Comment:
This seems unnecessarily complicated to me. Do we really need all these level of lazy? In particular, it seems to me that we never force only one of `files` or `unfiltered_sub_dirs`, which makes me thing that it is not necessary to return them as separate lazy values.

Reply:
To read the project file you only need files actually. Here I disagree that laziness complicates things as its just a flat list of bindings.  But since it doesn't buy us all that much I can get rid of it.

Reply:
In practice, I don't think we can be in a situation where we would read the dune-project file and not recurse through the sub-directories. So getting rid of it seems best to me, at least it simplifies the code

Reply:
Yeah, but there are certainly cases where we don't need to sort all the files. For big directories, that might not be so small.